src/VisualStudio/Core/Def/RoslynPackage.cs

```diff
@@ -208,5 +211,71 @@ private void UnregisterRuleSetEventHandler()
                 _ruleSetEventHandler = null;
             }
         }
+
+        private void TrackBulkFileOperations()
+        {
+            // we will pause whatever ambient work loads we have that are tied to IGlobalOperationNotificationService
+            // such as solution crawler, pre-emptive remote host synchronization and etc. any background work users didn't
+            // explicitly asked for.
+            //
+            // this should give all resources to BulkFileOperation. we do same for things like build, 
+            // debugging, wait dialog and etc. BulkFileOperation is used for things like git branch switching and etc.
+            var globalNotificationService = _workspace.Services.GetService<IGlobalOperationNotificationService>();
+
+            // BulkFileOperation can't have nested events. there will be ever only 1 events (Begin/End)
+            // so we only need simple tracking.
+            object gate = new object();
+            GlobalOperationRegistration localRegistration = null;
+
+            BulkFileOperation.Begin += (s, a) =>
+            {
+                StartBulkFileOperationNotification();
+            };
+
+            BulkFileOperation.End += (s, a) =>
```

Comment:
What is the threading expectation here -- we can get bulk operations from any thread? What about icky cases like:

1. Register for BulkFileOperation.Begin
1. An operation starts, so we start the global operation.
1. An operation ends, but we haven't subscribed yet.
1. Register for BulkFileOperation.End.